{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "32592fc1",
   "metadata": {},
   "source": [
    "This is a notebook for the regesion model, 2.iv "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f250ebb7",
   "metadata": {},
   "source": [
    "###  General Idea\n",
    "\n",
    "> **“As a group, how much of a gene’s expression variation across cell types can be **explained by** the accessibility of its candidate CREs?”**\n",
    "\n",
    "Concretely, for each gene you’ll:\n",
    "\n",
    "1. **Gather predictors**  \n",
    "   – All your ATAC-seq peaks (CREs) linked to that gene (e.g. ±20 kb around its TSS or however you defined your peak→gene map).  \n",
    "   – For each of these peaks, you already have a vector of accessibility values across N cell types.\n",
    "\n",
    "2. **Gather response**  \n",
    "   – The same gene’s expression levels across those N cell types.\n",
    "\n",
    "3. **Fit a linear model**  \n",
    "   –  where  \n",
    "     - **Y** is the N-vector of gene expression,  \n",
    "     - **X** is the N×P matrix of peak-accessibility (one column per CRE),  \n",
    "     - **β** are the coefficients.  \n",
    "     – Compute **R²** to see “what fraction of variance in Y is captured by X.”\n",
    "\n",
    "4. **Inspect coefficients**  \n",
    "   – Which CREs get large positive (activating) or negative (repressing) **β**?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d6a543f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ImmGenATAC1219.peakID', 'chrom', 'Summit', 'mm10.60way.phastCons_scores', '_-log10_bestPvalue', 'Included.in.systematic.analysis', 'TSS', 'genes.within.100Kb', 'preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th', 'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th', 'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.4.Sp.aCD3+CD40.18hr', 'T.8.Nve.Sp', 'Treg.4.25hi.Sp', 'Treg.4.FP3+.Nrplo.Co', 'T8.TN.P14.Sp', 'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp', 'T8.IEL.LCMV.d7.SI', 'T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp', 'NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d', 'LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ImmGenATAC1219.peakID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>Summit</th>\n",
       "      <th>mm10.60way.phastCons_scores</th>\n",
       "      <th>_-log10_bestPvalue</th>\n",
       "      <th>Included.in.systematic.analysis</th>\n",
       "      <th>TSS</th>\n",
       "      <th>genes.within.100Kb</th>\n",
       "      <th>preT.DN1.Th</th>\n",
       "      <th>preT.DN2a.Th</th>\n",
       "      <th>...</th>\n",
       "      <th>T8.Tcm.LCMV.d180.Sp</th>\n",
       "      <th>T8.Tem.LCMV.d180.Sp</th>\n",
       "      <th>NKT.Sp</th>\n",
       "      <th>NKT.Sp.LPS.3hr</th>\n",
       "      <th>NKT.Sp.LPS.18hr</th>\n",
       "      <th>NKT.Sp.LPS.3d</th>\n",
       "      <th>LTHSC.34-.BM</th>\n",
       "      <th>LTHSC.34+.BM</th>\n",
       "      <th>STHSC.150-.BM</th>\n",
       "      <th>MPP4.135+.BM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ImmGenATAC1219.peak_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3020786</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>...</td>\n",
       "      <td>0.71</td>\n",
       "      <td>2.78</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ImmGenATAC1219.peak_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3087226</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.77</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>3.56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2.12</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.41</td>\n",
       "      <td>1.64</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ImmGenATAC1219.peak_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3120109</td>\n",
       "      <td>0.07</td>\n",
       "      <td>10.80</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.77</td>\n",
       "      <td>...</td>\n",
       "      <td>1.32</td>\n",
       "      <td>2.79</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.72</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.36</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ImmGenATAC1219.peak_4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3121485</td>\n",
       "      <td>0.15</td>\n",
       "      <td>3.02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.11</td>\n",
       "      <td>...</td>\n",
       "      <td>3.41</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.86</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ImmGenATAC1219.peak_5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3372787</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.84</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.59</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ImmGenATAC1219.peakID chrom   Summit  mm10.60way.phastCons_scores  \\\n",
       "0  ImmGenATAC1219.peak_1  chr1  3020786                         0.00   \n",
       "1  ImmGenATAC1219.peak_2  chr1  3087226                         0.00   \n",
       "2  ImmGenATAC1219.peak_3  chr1  3120109                         0.07   \n",
       "3  ImmGenATAC1219.peak_4  chr1  3121485                         0.15   \n",
       "4  ImmGenATAC1219.peak_5  chr1  3372787                         0.03   \n",
       "\n",
       "   _-log10_bestPvalue  Included.in.systematic.analysis  TSS  \\\n",
       "0                0.56                              NaN  NaN   \n",
       "1                0.50                              NaN  NaN   \n",
       "2               10.80                              1.0  NaN   \n",
       "3                3.02                              1.0  NaN   \n",
       "4                1.31                              NaN  NaN   \n",
       "\n",
       "  genes.within.100Kb  preT.DN1.Th  preT.DN2a.Th  ...  T8.Tcm.LCMV.d180.Sp  \\\n",
       "0                NaN         0.11          0.77  ...                 0.71   \n",
       "1                NaN         0.77          0.77  ...                 0.94   \n",
       "2                NaN         0.40          0.77  ...                 1.32   \n",
       "3                NaN         0.46          0.11  ...                 3.41   \n",
       "4                NaN         0.77          0.84  ...                 0.94   \n",
       "\n",
       "   T8.Tem.LCMV.d180.Sp  NKT.Sp  NKT.Sp.LPS.3hr  NKT.Sp.LPS.18hr  \\\n",
       "0                 2.78    0.56            0.11             0.64   \n",
       "1                 3.56    0.98            2.12             0.12   \n",
       "2                 2.79    0.56            1.72             0.12   \n",
       "3                 0.11    2.86            0.11             0.64   \n",
       "4                 0.11    1.59            0.51             0.12   \n",
       "\n",
       "   NKT.Sp.LPS.3d  LTHSC.34-.BM  LTHSC.34+.BM  STHSC.150-.BM  MPP4.135+.BM  \n",
       "0           1.14          0.41          0.71           0.90          0.11  \n",
       "1           0.11          0.41          1.64           0.90          0.83  \n",
       "2           0.11          2.36          0.10           0.90          0.11  \n",
       "3           2.20          0.41          0.10           0.11          0.11  \n",
       "4           0.53          0.41          0.10           0.11          0.11  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 1) Load the full table\n",
    "atac = pd.read_csv(\"ATAC-seq/filtered_ATAC_abT_Tact_Stem.csv\", index_col=False)\n",
    "\n",
    "# 2) See exactly which columns it contains\n",
    "print(atac.columns.tolist())\n",
    "\n",
    "# (Optional) peek at the first few rows\n",
    "atac.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "23910af1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['peak_id', 'preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th', 'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th', 'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.4.Sp.aCD3+CD40.18hr', 'T.8.Nve.Sp', 'Treg.4.25hi.Sp', 'Treg.4.FP3+.Nrplo.Co', 'T8.TN.P14.Sp', 'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp', 'T8.IEL.LCMV.d7.SI', 'T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp', 'NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d', 'LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_id</th>\n",
       "      <th>preT.DN1.Th</th>\n",
       "      <th>preT.DN2a.Th</th>\n",
       "      <th>preT.DN2b.Th</th>\n",
       "      <th>preT.DN3.Th</th>\n",
       "      <th>T.DN4.Th</th>\n",
       "      <th>T.ISP.Th</th>\n",
       "      <th>T.DP.Th</th>\n",
       "      <th>T.4.Th</th>\n",
       "      <th>T.8.Th</th>\n",
       "      <th>...</th>\n",
       "      <th>T8.Tcm.LCMV.d180.Sp</th>\n",
       "      <th>T8.Tem.LCMV.d180.Sp</th>\n",
       "      <th>NKT.Sp</th>\n",
       "      <th>NKT.Sp.LPS.3hr</th>\n",
       "      <th>NKT.Sp.LPS.18hr</th>\n",
       "      <th>NKT.Sp.LPS.3d</th>\n",
       "      <th>LTHSC.34-.BM</th>\n",
       "      <th>LTHSC.34+.BM</th>\n",
       "      <th>STHSC.150-.BM</th>\n",
       "      <th>MPP4.135+.BM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ImmGenATAC1219.peak_1</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.09</td>\n",
       "      <td>...</td>\n",
       "      <td>0.71</td>\n",
       "      <td>2.78</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ImmGenATAC1219.peak_2</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.47</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.02</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.53</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>3.56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2.12</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.41</td>\n",
       "      <td>1.64</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ImmGenATAC1219.peak_3</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.53</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>...</td>\n",
       "      <td>1.32</td>\n",
       "      <td>2.79</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.72</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.36</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ImmGenATAC1219.peak_4</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.17</td>\n",
       "      <td>2.15</td>\n",
       "      <td>0.53</td>\n",
       "      <td>...</td>\n",
       "      <td>3.41</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.86</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ImmGenATAC1219.peak_5</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.84</td>\n",
       "      <td>1.34</td>\n",
       "      <td>1.69</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.54</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.59</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 peak_id  preT.DN1.Th  preT.DN2a.Th  preT.DN2b.Th  \\\n",
       "0  ImmGenATAC1219.peak_1         0.11          0.77          0.51   \n",
       "1  ImmGenATAC1219.peak_2         0.77          0.77          0.47   \n",
       "2  ImmGenATAC1219.peak_3         0.40          0.77          0.11   \n",
       "3  ImmGenATAC1219.peak_4         0.46          0.11          0.47   \n",
       "4  ImmGenATAC1219.peak_5         0.77          0.84          1.34   \n",
       "\n",
       "   preT.DN3.Th  T.DN4.Th  T.ISP.Th  T.DP.Th  T.4.Th  T.8.Th  ...  \\\n",
       "0         0.49      0.58      0.11     0.67    0.11    1.09  ...   \n",
       "1         1.56      1.02      0.11     0.73    0.91    0.53  ...   \n",
       "2         0.10      0.35      0.53     1.08    0.11    0.11  ...   \n",
       "3         0.10      0.58      0.11     0.17    2.15    0.53  ...   \n",
       "4         1.69      0.34      0.94     0.40    0.39    0.54  ...   \n",
       "\n",
       "   T8.Tcm.LCMV.d180.Sp  T8.Tem.LCMV.d180.Sp  NKT.Sp  NKT.Sp.LPS.3hr  \\\n",
       "0                 0.71                 2.78    0.56            0.11   \n",
       "1                 0.94                 3.56    0.98            2.12   \n",
       "2                 1.32                 2.79    0.56            1.72   \n",
       "3                 3.41                 0.11    2.86            0.11   \n",
       "4                 0.94                 0.11    1.59            0.51   \n",
       "\n",
       "   NKT.Sp.LPS.18hr  NKT.Sp.LPS.3d  LTHSC.34-.BM  LTHSC.34+.BM  STHSC.150-.BM  \\\n",
       "0             0.64           1.14          0.41          0.71           0.90   \n",
       "1             0.12           0.11          0.41          1.64           0.90   \n",
       "2             0.12           0.11          2.36          0.10           0.90   \n",
       "3             0.64           2.20          0.41          0.10           0.11   \n",
       "4             0.12           0.53          0.41          0.10           0.11   \n",
       "\n",
       "   MPP4.135+.BM  \n",
       "0          0.11  \n",
       "1          0.83  \n",
       "2          0.11  \n",
       "3          0.11  \n",
       "4          0.11  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 1) load and rename the peak‐ID column\n",
    "atac = pd.read_csv(\"ATAC-seq/filtered_ATAC_abT_Tact_Stem.csv\")\n",
    "atac = atac.rename(columns={\"ImmGenATAC1219.peakID\": \"peak_id\"})\n",
    "\n",
    "# 2) list exactly the metadata columns you want to remove\n",
    "to_drop = [\n",
    "    \"chrom\",\n",
    "    \"Summit\",\n",
    "    \"mm10.60way.phastCons_scores\",\n",
    "    \"_-log10_bestPvalue\",\n",
    "    \"Included.in.systematic.analysis\",\n",
    "    \"TSS\",\n",
    "    \"genes.within.100Kb\",\n",
    "]\n",
    "\n",
    "# 3) drop them, keeping peak_id + everything else\n",
    "atac_sub = atac.drop(columns=to_drop)\n",
    "\n",
    "# 4) confirm\n",
    "print(atac_sub.columns.tolist())\n",
    "atac_sub.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0a569848",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATAC shape: (512595, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preT.DN1.Th</th>\n",
       "      <th>preT.DN2a.Th</th>\n",
       "      <th>preT.DN2b.Th</th>\n",
       "      <th>preT.DN3.Th</th>\n",
       "      <th>T.DN4.Th</th>\n",
       "      <th>T.ISP.Th</th>\n",
       "      <th>T.DP.Th</th>\n",
       "      <th>T.4.Th</th>\n",
       "      <th>T.8.Th</th>\n",
       "      <th>T.4.Nve.Sp</th>\n",
       "      <th>...</th>\n",
       "      <th>T8.Tcm.LCMV.d180.Sp</th>\n",
       "      <th>T8.Tem.LCMV.d180.Sp</th>\n",
       "      <th>NKT.Sp</th>\n",
       "      <th>NKT.Sp.LPS.3hr</th>\n",
       "      <th>NKT.Sp.LPS.18hr</th>\n",
       "      <th>NKT.Sp.LPS.3d</th>\n",
       "      <th>LTHSC.34-.BM</th>\n",
       "      <th>LTHSC.34+.BM</th>\n",
       "      <th>STHSC.150-.BM</th>\n",
       "      <th>MPP4.135+.BM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peak_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_1</th>\n",
       "      <td>0.11</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.09</td>\n",
       "      <td>3.23</td>\n",
       "      <td>...</td>\n",
       "      <td>0.71</td>\n",
       "      <td>2.78</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_2</th>\n",
       "      <td>0.77</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.47</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.02</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.84</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>3.56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2.12</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.41</td>\n",
       "      <td>1.64</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_3</th>\n",
       "      <td>0.40</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.53</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.04</td>\n",
       "      <td>...</td>\n",
       "      <td>1.32</td>\n",
       "      <td>2.79</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.72</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.36</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_4</th>\n",
       "      <td>0.46</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.17</td>\n",
       "      <td>2.15</td>\n",
       "      <td>0.53</td>\n",
       "      <td>1.25</td>\n",
       "      <td>...</td>\n",
       "      <td>3.41</td>\n",
       "      <td>0.11</td>\n",
       "      <td>2.86</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_5</th>\n",
       "      <td>0.77</td>\n",
       "      <td>0.84</td>\n",
       "      <td>1.34</td>\n",
       "      <td>1.69</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.54</td>\n",
       "      <td>1.89</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.59</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       preT.DN1.Th  preT.DN2a.Th  preT.DN2b.Th  preT.DN3.Th  \\\n",
       "peak_id                                                                       \n",
       "ImmGenATAC1219.peak_1         0.11          0.77          0.51         0.49   \n",
       "ImmGenATAC1219.peak_2         0.77          0.77          0.47         1.56   \n",
       "ImmGenATAC1219.peak_3         0.40          0.77          0.11         0.10   \n",
       "ImmGenATAC1219.peak_4         0.46          0.11          0.47         0.10   \n",
       "ImmGenATAC1219.peak_5         0.77          0.84          1.34         1.69   \n",
       "\n",
       "                       T.DN4.Th  T.ISP.Th  T.DP.Th  T.4.Th  T.8.Th  \\\n",
       "peak_id                                                              \n",
       "ImmGenATAC1219.peak_1      0.58      0.11     0.67    0.11    1.09   \n",
       "ImmGenATAC1219.peak_2      1.02      0.11     0.73    0.91    0.53   \n",
       "ImmGenATAC1219.peak_3      0.35      0.53     1.08    0.11    0.11   \n",
       "ImmGenATAC1219.peak_4      0.58      0.11     0.17    2.15    0.53   \n",
       "ImmGenATAC1219.peak_5      0.34      0.94     0.40    0.39    0.54   \n",
       "\n",
       "                       T.4.Nve.Sp  ...  T8.Tcm.LCMV.d180.Sp  \\\n",
       "peak_id                            ...                        \n",
       "ImmGenATAC1219.peak_1        3.23  ...                 0.71   \n",
       "ImmGenATAC1219.peak_2        0.84  ...                 0.94   \n",
       "ImmGenATAC1219.peak_3        2.04  ...                 1.32   \n",
       "ImmGenATAC1219.peak_4        1.25  ...                 3.41   \n",
       "ImmGenATAC1219.peak_5        1.89  ...                 0.94   \n",
       "\n",
       "                       T8.Tem.LCMV.d180.Sp  NKT.Sp  NKT.Sp.LPS.3hr  \\\n",
       "peak_id                                                              \n",
       "ImmGenATAC1219.peak_1                 2.78    0.56            0.11   \n",
       "ImmGenATAC1219.peak_2                 3.56    0.98            2.12   \n",
       "ImmGenATAC1219.peak_3                 2.79    0.56            1.72   \n",
       "ImmGenATAC1219.peak_4                 0.11    2.86            0.11   \n",
       "ImmGenATAC1219.peak_5                 0.11    1.59            0.51   \n",
       "\n",
       "                       NKT.Sp.LPS.18hr  NKT.Sp.LPS.3d  LTHSC.34-.BM  \\\n",
       "peak_id                                                               \n",
       "ImmGenATAC1219.peak_1             0.64           1.14          0.41   \n",
       "ImmGenATAC1219.peak_2             0.12           0.11          0.41   \n",
       "ImmGenATAC1219.peak_3             0.12           0.11          2.36   \n",
       "ImmGenATAC1219.peak_4             0.64           2.20          0.41   \n",
       "ImmGenATAC1219.peak_5             0.12           0.53          0.41   \n",
       "\n",
       "                       LTHSC.34+.BM  STHSC.150-.BM  MPP4.135+.BM  \n",
       "peak_id                                                           \n",
       "ImmGenATAC1219.peak_1          0.71           0.90          0.11  \n",
       "ImmGenATAC1219.peak_2          1.64           0.90          0.83  \n",
       "ImmGenATAC1219.peak_3          0.10           0.90          0.11  \n",
       "ImmGenATAC1219.peak_4          0.10           0.11          0.11  \n",
       "ImmGenATAC1219.peak_5          0.10           0.11          0.11  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 1) Load without an index\n",
    "atac = pd.read_csv(\"ATAC-seq/filtered_ATAC_abT_Tact_Stem.csv\")\n",
    "\n",
    "# 2) Rename your peak‐ID column\n",
    "atac = atac.rename(columns={\"ImmGenATAC1219.peakID\": \"peak_id\"})\n",
    "\n",
    "# 3) Drop any leftover metadata if you haven’t yet\n",
    "to_drop = [\n",
    "    \"chrom\", \"Summit\", \"mm10.60way.phastCons_scores\",\n",
    "    \"_-log10_bestPvalue\", \"Included.in.systematic.analysis\",\n",
    "    \"TSS\", \"genes.within.100Kb\",\n",
    "]\n",
    "atac_sub = atac.drop(columns=to_drop)\n",
    "\n",
    "# 4) Now set the index\n",
    "atac_sub = atac_sub.set_index(\"peak_id\")\n",
    "\n",
    "# 5) Quick check\n",
    "print(\"ATAC shape:\", atac_sub.shape)\n",
    "atac_sub.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c15a1ea7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17535, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preT.DN1.Th</th>\n",
       "      <th>preT.DN2a.Th</th>\n",
       "      <th>preT.DN2b.Th</th>\n",
       "      <th>preT.DN3.Th</th>\n",
       "      <th>T.DN4.Th</th>\n",
       "      <th>T.ISP.Th</th>\n",
       "      <th>T.DP.Th</th>\n",
       "      <th>T.4.Th</th>\n",
       "      <th>T.8.Th</th>\n",
       "      <th>T.4.Nve.Sp</th>\n",
       "      <th>...</th>\n",
       "      <th>T8.Tcm.LCMV.d180.Sp</th>\n",
       "      <th>T8.Tem.LCMV.d180.Sp</th>\n",
       "      <th>NKT.Sp</th>\n",
       "      <th>NKT.Sp.LPS.3hr</th>\n",
       "      <th>NKT.Sp.LPS.18hr</th>\n",
       "      <th>NKT.Sp.LPS.3d</th>\n",
       "      <th>LTHSC.34-.BM</th>\n",
       "      <th>LTHSC.34+.BM</th>\n",
       "      <th>STHSC.150-.BM</th>\n",
       "      <th>MPP4.135+.BM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>1.022363</td>\n",
       "      <td>1.389747</td>\n",
       "      <td>1.024819</td>\n",
       "      <td>1.024482</td>\n",
       "      <td>1.026430</td>\n",
       "      <td>1.026217</td>\n",
       "      <td>3.010920</td>\n",
       "      <td>1.024462</td>\n",
       "      <td>1.024819</td>\n",
       "      <td>2.726341</td>\n",
       "      <td>...</td>\n",
       "      <td>1.025833</td>\n",
       "      <td>1.024819</td>\n",
       "      <td>1.385805</td>\n",
       "      <td>1.025833</td>\n",
       "      <td>1.575395</td>\n",
       "      <td>1.024819</td>\n",
       "      <td>1.096732</td>\n",
       "      <td>1.096732</td>\n",
       "      <td>1.021750</td>\n",
       "      <td>1.021812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>162.641117</td>\n",
       "      <td>206.945221</td>\n",
       "      <td>209.187788</td>\n",
       "      <td>198.421365</td>\n",
       "      <td>215.056475</td>\n",
       "      <td>225.565360</td>\n",
       "      <td>73.904647</td>\n",
       "      <td>138.841383</td>\n",
       "      <td>139.863904</td>\n",
       "      <td>168.924363</td>\n",
       "      <td>...</td>\n",
       "      <td>206.241084</td>\n",
       "      <td>205.309922</td>\n",
       "      <td>165.690720</td>\n",
       "      <td>133.234920</td>\n",
       "      <td>127.894194</td>\n",
       "      <td>195.147548</td>\n",
       "      <td>206.053987</td>\n",
       "      <td>246.105317</td>\n",
       "      <td>192.424636</td>\n",
       "      <td>204.298358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>68.070719</td>\n",
       "      <td>82.468806</td>\n",
       "      <td>89.769337</td>\n",
       "      <td>57.661619</td>\n",
       "      <td>76.399214</td>\n",
       "      <td>84.671456</td>\n",
       "      <td>32.828651</td>\n",
       "      <td>27.207241</td>\n",
       "      <td>36.169759</td>\n",
       "      <td>32.753248</td>\n",
       "      <td>...</td>\n",
       "      <td>36.105700</td>\n",
       "      <td>34.348965</td>\n",
       "      <td>25.168975</td>\n",
       "      <td>33.305724</td>\n",
       "      <td>29.284365</td>\n",
       "      <td>33.322384</td>\n",
       "      <td>78.272059</td>\n",
       "      <td>78.837030</td>\n",
       "      <td>68.844751</td>\n",
       "      <td>76.418169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>15.450717</td>\n",
       "      <td>13.573968</td>\n",
       "      <td>14.427620</td>\n",
       "      <td>8.249482</td>\n",
       "      <td>1.683173</td>\n",
       "      <td>4.001953</td>\n",
       "      <td>5.595954</td>\n",
       "      <td>6.367369</td>\n",
       "      <td>6.505833</td>\n",
       "      <td>8.262234</td>\n",
       "      <td>...</td>\n",
       "      <td>8.645607</td>\n",
       "      <td>7.268431</td>\n",
       "      <td>3.840215</td>\n",
       "      <td>1.025833</td>\n",
       "      <td>6.283540</td>\n",
       "      <td>12.791348</td>\n",
       "      <td>8.577159</td>\n",
       "      <td>16.791386</td>\n",
       "      <td>15.511549</td>\n",
       "      <td>16.947354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <td>160.246297</td>\n",
       "      <td>125.475307</td>\n",
       "      <td>155.928005</td>\n",
       "      <td>120.692893</td>\n",
       "      <td>118.433597</td>\n",
       "      <td>149.630866</td>\n",
       "      <td>92.040668</td>\n",
       "      <td>76.781112</td>\n",
       "      <td>87.529814</td>\n",
       "      <td>86.523573</td>\n",
       "      <td>...</td>\n",
       "      <td>87.608325</td>\n",
       "      <td>56.128251</td>\n",
       "      <td>109.175415</td>\n",
       "      <td>91.992319</td>\n",
       "      <td>102.035627</td>\n",
       "      <td>108.414405</td>\n",
       "      <td>168.645852</td>\n",
       "      <td>157.926022</td>\n",
       "      <td>155.941641</td>\n",
       "      <td>186.261464</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               preT.DN1.Th  preT.DN2a.Th  preT.DN2b.Th  preT.DN3.Th  \\\n",
       "gene_symbol                                                           \n",
       "0610005C13Rik     1.022363      1.389747      1.024819     1.024482   \n",
       "0610007P14Rik   162.641117    206.945221    209.187788   198.421365   \n",
       "0610009B22Rik    68.070719     82.468806     89.769337    57.661619   \n",
       "0610009L18Rik    15.450717     13.573968     14.427620     8.249482   \n",
       "0610009O20Rik   160.246297    125.475307    155.928005   120.692893   \n",
       "\n",
       "                 T.DN4.Th    T.ISP.Th    T.DP.Th      T.4.Th      T.8.Th  \\\n",
       "gene_symbol                                                                \n",
       "0610005C13Rik    1.026430    1.026217   3.010920    1.024462    1.024819   \n",
       "0610007P14Rik  215.056475  225.565360  73.904647  138.841383  139.863904   \n",
       "0610009B22Rik   76.399214   84.671456  32.828651   27.207241   36.169759   \n",
       "0610009L18Rik    1.683173    4.001953   5.595954    6.367369    6.505833   \n",
       "0610009O20Rik  118.433597  149.630866  92.040668   76.781112   87.529814   \n",
       "\n",
       "               T.4.Nve.Sp  ...  T8.Tcm.LCMV.d180.Sp  T8.Tem.LCMV.d180.Sp  \\\n",
       "gene_symbol                ...                                             \n",
       "0610005C13Rik    2.726341  ...             1.025833             1.024819   \n",
       "0610007P14Rik  168.924363  ...           206.241084           205.309922   \n",
       "0610009B22Rik   32.753248  ...            36.105700            34.348965   \n",
       "0610009L18Rik    8.262234  ...             8.645607             7.268431   \n",
       "0610009O20Rik   86.523573  ...            87.608325            56.128251   \n",
       "\n",
       "                   NKT.Sp  NKT.Sp.LPS.3hr  NKT.Sp.LPS.18hr  NKT.Sp.LPS.3d  \\\n",
       "gene_symbol                                                                 \n",
       "0610005C13Rik    1.385805        1.025833         1.575395       1.024819   \n",
       "0610007P14Rik  165.690720      133.234920       127.894194     195.147548   \n",
       "0610009B22Rik   25.168975       33.305724        29.284365      33.322384   \n",
       "0610009L18Rik    3.840215        1.025833         6.283540      12.791348   \n",
       "0610009O20Rik  109.175415       91.992319       102.035627     108.414405   \n",
       "\n",
       "               LTHSC.34-.BM  LTHSC.34+.BM  STHSC.150-.BM  MPP4.135+.BM  \n",
       "gene_symbol                                                             \n",
       "0610005C13Rik      1.096732      1.096732       1.021750      1.021812  \n",
       "0610007P14Rik    206.053987    246.105317     192.424636    204.298358  \n",
       "0610009B22Rik     78.272059     78.837030      68.844751     76.418169  \n",
       "0610009L18Rik      8.577159     16.791386      15.511549     16.947354  \n",
       "0610009O20Rik    168.645852    157.926022     155.941641    186.261464  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 1) Load without setting an index\n",
    "rna = pd.read_csv(\"RNA-seq/filtered_RNA_abT_Tact_Stem.csv\")\n",
    "\n",
    "# Rename and index\n",
    "rna = (\n",
    "    rna\n",
    "    .rename(columns={\"Unnamed: 0\": \"gene_symbol\"})\n",
    "    .set_index(\"gene_symbol\")\n",
    ")\n",
    "\n",
    "# Confirm\n",
    "print(rna.shape)\n",
    "rna.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c2b5733f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATAC shape: (512595, 29)\n",
      "RNA  shape: (17535, 29)\n",
      "Annot shape: (14859136, 10)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# 3) Load your peak→gene annotation\n",
    "annot = pd.read_csv(\"peak_to_gene_annotated.csv\")\n",
    "\n",
    "# Quick checks\n",
    "print(\"ATAC shape:\", atac_sub.shape)\n",
    "print(\"RNA  shape:\",  rna.shape)\n",
    "print(\"Annot shape:\", annot.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ba54dd44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0610005C13Rik R² = 0.183\n",
      "Top 5 CREs by |β|:\n",
      " peak_id\n",
      "ImmGenATAC1219.peak_427239    5.747318e+12\n",
      "ImmGenATAC1219.peak_427235    3.944051e+12\n",
      "ImmGenATAC1219.peak_427236    3.125048e+12\n",
      "ImmGenATAC1219.peak_427237    2.711480e+12\n",
      "ImmGenATAC1219.peak_427240    2.634922e+12\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "from sklearn.linear_model import LinearRegression\n",
    "import pandas as pd\n",
    "\n",
    "# 1) Your gene\n",
    "gene = \"0610005C13Rik\"\n",
    "\n",
    "# 2) Response vector Y\n",
    "Y = rna.loc[gene]\n",
    "\n",
    "# 3) Predictor matrix X: all peaks linked to that gene\n",
    "peaks = annot.loc[annot.gene_symbol == gene, \"peak_id\"].tolist()\n",
    "X = atac_sub.loc[peaks].T\n",
    "\n",
    "# 4) Fit and report R²\n",
    "model = LinearRegression().fit(X, Y)\n",
    "print(f\"{gene} R² = {model.score(X, Y):.3f}\")\n",
    "\n",
    "# 5) Top 5 strongest CREs by |β|\n",
    "coefs = pd.Series(model.coef_, index=X.columns).abs().nlargest(5)\n",
    "print(\"Top 5 CREs by |β|:\\n\", coefs)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "REAL_FORKED_Bioinfo_project",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
